# ANSWER 1

### 1.1 Importing all required libraries

In [2]:
import re
import numpy as np
import pandas as pd
from sklearn import *
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

C:\Users\hp\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
C:\Users\hp\Anaconda3\lib\site-packages\sklearn\grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)
C:\Users\hp\Anaconda3\lib\site-packages\sklearn\learning_curve.py:22: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the functions are moved. This module will be removed in 0.20
  DeprecationWarning)


### 1.2 Importing DataSets

In [3]:
test_df = pd.read_csv('test.csv')
train_df = pd.read_csv('train.csv')

In [4]:
test_df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [41]:
print('The number of samples into the test data i.e. test_df is {}.'.format(test_df.shape[0]))

The number of samples into the test data i.e. test_df is 418.


## NOTE : In  test_df we don't have any Column for Survived.
#### We have make the use of head function because we wants to know that  the test_df(test.csv) starts from which  PassengerId  i.e 892 and we also also wants to know that where its ends thats why we have make the use of tail function to finding its ending i.e. 417

In [45]:
test_df.tail()

PassengerId  Pclass                          Name     Sex   Age  SibSp  \
413         1305       3            Spector, Mr. Woolf    male   NaN      0   
414         1306       1  Oliva y Ocana, Dona. Fermina  female  39.0      0   
415         1307       3  Saether, Mr. Simon Sivertsen    male  38.5      0   
416         1308       3           Ware, Mr. Frederick    male   NaN      0   
417         1309       3      Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  
413      0           A.5. 3236    8.0500   NaN        S  
414      0            PC 17758  108.9000  C105        C  
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0              359309    8.0500   NaN        S  
417      1                2668   22.3583   NaN        C

In [10]:
train_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# NOTE : In train_df we have a column for Survived.
#### We have make the use of head function because we wants to know that  the train_df(train.csv) starts from which  PassengerId  i.e 1 and we also also wants to know that where its ends thats why we have make the use of tail function to finding its ending i.e. 891

In [11]:
train_df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

### Concatenating the two DataSets

In [12]:
concated_dataframe = [train_df,test_df]

In [13]:
result = pd.concat(concated_dataframe)

C:\Users\hp\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=True'.

To retain the current behavior and silence the warning, pass sort=False

  """Entry point for launching an IPython kernel.


### Display first 5 rows of concatenated Dataset

In [14]:
result.head()

Age Cabin Embarked     Fare  \
0  22.0   NaN        S   7.2500   
1  38.0   C85        C  71.2833   
2  26.0   NaN        S   7.9250   
3  35.0  C123        S  53.1000   
4  35.0   NaN        S   8.0500   

                                                Name  Parch  PassengerId  \
0                            Braund, Mr. Owen Harris      0            1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0            2   
2                             Heikkinen, Miss. Laina      0            3   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      0            4   
4                           Allen, Mr. William Henry      0            5   

   Pclass     Sex  SibSp  Survived            Ticket  
0       3    male      1       0.0         A/5 21171  
1       1  female      1       1.0          PC 17599  
2       3  female      0       1.0  STON/O2. 3101282  
3       1  female      1       1.0            113803  
4       3    male      0       0.0            373450

### from the above , first row there's a missing value in Cabin Column so we will now handle the missing data. For that you have to proceed to answer 2


### the Data given below we have concluded that there are many missing values in Cabin Column.

In [15]:
result.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1309 entries, 0 to 417
Data columns (total 12 columns):
Age            1046 non-null float64
Cabin          295 non-null object
Embarked       1307 non-null object
Fare           1308 non-null float64
Name           1309 non-null object
Parch          1309 non-null int64
PassengerId    1309 non-null int64
Pclass         1309 non-null int64
Sex            1309 non-null object
SibSp          1309 non-null int64
Survived       891 non-null float64
Ticket         1309 non-null object
dtypes: float64(3), int64(4), object(5)
memory usage: 132.9+ KB


# ANSWER 2

### The function given below tells us about that how many missing values are there in a each row.
### In zeroth row there is only 1 missing value.
### In first row there is no Missing value all Columns of that row completely filled.
### and so on.

In [51]:
result.apply(lambda x: x.count(), axis=1)

0      11
1      12
2      11
3      12
4      11
5      10
6      12
7      11
8      11
9      11
10     12
11     12
12     11
13     11
14     11
15     11
16     11
17     10
18     11
19     10
20     11
21     12
22     11
23     12
24     11
25     11
26     10
27     12
28     10
29     10
       ..
388    10
389    10
390    11
391    11
392    10
393    10
394    10
395    11
396    10
397    11
398    10
399    10
400    11
401    10
402    10
403    10
404    11
405    11
406    10
407    11
408     9
409    10
410     9
411    11
412    10
413     9
414    11
415    10
416     9
417     9
Length: 1309, dtype: int64

## Note: there is no target variable into test data (i.e. "Survival" column is missing), so the goal is to predict this target using different machine learning algorithms such as logistic regression.

In [16]:
result['Survived'].value_counts()

0.0    549
1.0    342
Name: Survived, dtype: int64

In [17]:
result.isnull().sum()

Age             263
Cabin          1014
Embarked          2
Fare              1
Name              0
Parch             0
PassengerId       0
Pclass            0
Sex               0
SibSp             0
Survived        418
Ticket            0
dtype: int64

In [18]:
result = result.dropna(subset=['Survived'])

In [19]:
result.isnull().sum()

Age            177
Cabin          687
Embarked         2
Fare             0
Name             0
Parch            0
PassengerId      0
Pclass           0
Sex              0
SibSp            0
Survived         0
Ticket           0
dtype: int64

# Answer 3

In [20]:
i=0
result['Has_Cabin']=""
for x in result['Cabin'].isnull():
  if x==True:
    result.loc[i,'Has_Cabin']=0
  else:
    result.loc[i,'Has_Cabin']=1
  i=i+1
result['Has_Cabin'].value_counts()

0    687
1    204
Name: Has_Cabin, dtype: int64

# Answer 4


In [21]:
result['Title']=result['Name'].apply(lambda x: re.search('([A-Z][a-z]+)\.', x))
result['Title']=result['Title'].apply(lambda x: x.group())
print(result['Title'].value_counts())

Mr.          517
Miss.        182
Mrs.         125
Master.       40
Dr.            7
Rev.           6
Col.           2
Major.         2
Mlle.          2
Jonkheer.      1
Countess.      1
Mme.           1
Don.           1
Lady.          1
Ms.            1
Capt.          1
Sir.           1
Name: Title, dtype: int64


In [22]:
result['Title'] = result['Title'].replace(['Don.', 'Dona.', 'Rev.', 'Dr.', 'Major.', 'Lady.', 'Sir.', 'Col.', 'Capt.', 'Countess.', 'Jonkheer.','Mme.','Mlle.','Ms.'],'Special')

In [23]:
result['Title'].value_counts()

Mr.        517
Miss.      182
Mrs.       125
Master.     40
Special     27
Name: Title, dtype: int64

# Answer 5

In [24]:
result['Age']=result['Age'].fillna(result['Age'].median())

In [25]:
result['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [26]:
result['Embarked'] = result['Embarked'].fillna('S')

# Answer 6


In [27]:
result['Categorical_age'] = pd.qcut(result.Age, q=4, labels=False )
result['Categorical_fare']= pd.qcut(result.Fare, q=4, labels=False)

# Answer 7

In [28]:
result.drop(['Cabin','Name','PassengerId','Ticket','Age','Fare','SibSp','Parch'],axis=1,inplace=True)

# Answer 8

In [29]:
result = pd.get_dummies(result)
result.head()

Pclass  Survived  Has_Cabin  Categorical_age  Categorical_fare  Embarked_C  \
0       3       0.0          0                0                 0           0   
1       1       1.0          1                3                 3           1   
2       3       1.0          0                1                 1           0   
3       1       1.0          1                2                 3           0   
4       3       0.0          0                2                 1           0   

   Embarked_Q  Embarked_S  Sex_female  Sex_male  Title_Master.  Title_Miss.  \
0           0           1           0         1              0            0   
1           0           0           1         0              0            0   
2           0           1           1         0              0            1   
3           0           1           1         0              0            0   
4           0           1           0         1              0            0   

   Title_Mr.  Title_Mrs.  Title_Special  
0          1           0              0  
1          0           1              0  
2          0           0              0  
3          0           1              0  
4          1           0              0

# Answer 9

In [30]:
x = result.drop(['Survived'], axis = 1)
y = result['Survived']

In [31]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3)

# Answer 10

In [32]:
model = LogisticRegression()

In [33]:
model.fit(x_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [34]:
predict = model.predict(x_test)

# Answer 11


In [36]:
accuracy_score = model.score(x_test, y_test)
print("The accuracy score is : {}".format(accuracy_score))

The accuracy score is : 0.8171641791044776
